pkg/apis/flowcontrol/types.go

```diff
@@ -0,0 +1,298 @@
+
+// FlowSchemaList is a list of FlowSchema objects.
+type FlowSchemaList struct {
+	metav1.TypeMeta
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta
+
+	// Items is a list of flow-schemas.
+	Items []FlowSchema
+}
+
+// FlowSchemaSpec describes how the flow-schema's specification looks like.
+type FlowSchemaSpec struct {
+	// MatchingPriority controls the ordering the flow-schemas in the cluster. Its value must be greater than zero and
+	// also unique. The primary sort is based on MatchingPriority, ordered highest to lowest (20 before 10).
+	MatchingPriority int32
+	// RequestPriorityRef can reference a RequestPriority in the cluster. If the reference cannot be resolved, the
+	// flow-schema will ignored and marked as invalid in its status.
+	RequestPriorityRef RequestPriorityRef
+	// Distinguisher defines how to classify requests into certain flows. If not empty, the flow's identifier is computed
+	// according to specification of the distinguisher.
+	// +optional
+	Distinguisher *FlowDistinguisher
+	// Rules holds all the RuleWithUser for this FlowSchema.
+	Rules []RuleWithUser
+}
+
+// FlowDistinguisher specifies the configuration of the distinguisher.
+type FlowDistinguisher struct {
+	// Source defines the source attribute we adopt to compute the identifier for the flow.
+	Source FlowDistinguisherSource
+}
+
+// FlowDistinguisherSource defines the input source of distinguishing flows.
+type FlowDistinguisherSource string
+
+// RequestPriorityRef contains information that points to the request-priority being used.
+type RequestPriorityRef struct {
+	// APIGroup is the group for the resource being referenced.
+	APIGroup string
+	// Kind is the type of resource being referenced.
+	Kind string
+	// Name is the name of resource being referenced
+	Name string
+}
+
+// RuleWithUser is a tuple of UserRule and RequestScopeRule.
+type RuleWithUser struct {
+	// User is is the user that this rule cares about.
+	User UserRule
+	// Rule is embedded, it describes other criteria of the rule, like
+	// APIGroups, APIVersions, Resources, etc.
+	Rule RequestScopeRule
+}
+
+// RuleWithUser is a tuple of User and Rule.
+type UserRule struct {
+	// Names is a list of user names the rule applies to. If '*' is present, the length of the slice must be one.
+	// Required.
+	Names []string
+	// Groups is a list of user group names the rule applies to. If '*' is present, the length of the slice must be one.
+	// Required.
+	Groups []string
+}
+
+type RequestScopeRule struct {
+	// APIGroups is the API groups the resources belong to. '*' is all groups.
+	// If '*' is present, the length of the slice must be one.
+	// Required.
+	APIGroups []string
+	// APIVersions is the API versions the resources belong to. '*' is all versions.
+	// If '*' is present, the length of the slice must be one.
+	// Required.
+	APIVersions []string
```

Comment:
My gut reaction is that users shouldn't need to worry about versions.  Inside of a group, the meaning of a resource should not change.  Why would we want to have limits different on a per version basis?

Reply:
good point. i can't think of any use-case where we discriminate requests per version. a tuple of group+resource+subresource+[nonResourceURL] seems sufficient for describing "what resource is the user requesting at"